File: app/src/commands/cheat/capital.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { Option } from "@tb-dev/utils";
import { invoke } from "@tauri-apps/api/core";
import type {
  BotId,
  CheatSetInfluenceRequest,
  Influence,
  PlayerId,
  PrecursorId,
  Ruler,
} from "@tsukilabs/nil-bindings";

export async function cheatSetInfluence(ruler: Option<Ruler>, influence: Influence) {
  const req: CheatSetInfluenceRequest = {
    world: NIL.world.getIdStrict(),
    ruler: ruler ?? null,
    influence,
  };

  await invoke("cheat_set_influence", { req });
}

export async function cheatSetBotInfluence(id: BotId, influence: Influence) {
  return cheatSetInfluence({ kind: "bot", id }, influence);
}

export async function cheatSetPlayerInfluence(id: PlayerId, influence: Influence) {
  return cheatSetInfluence({ kind: "player", id }, influence);
}

export async function cheatSetPrecursorInfluence(id: PrecursorId, influence: Influence) {
  return cheatSetInfluence({ kind: "precursor", id }, influence);
}

File: crates/nil-core/src/world/capital.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::ruler::Ruler;
use crate::world::World;
use tap::Pipe;

impl World {
  pub fn get_city_limit<R>(&self, ruler: R) -> Result<u32>
  where
    R: Into<Ruler>,
  {
    let ruler: Ruler = ruler.into();
    self
      .ruler(&ruler)?
      .influence()
      .city_limit()
      .pipe(Ok)
  }
}
